In [1]:
import pandas as pd
import numpy as np


Reading the data:

In [2]:
df = pd.read_csv("E:\Guvi\Projects\Inventory forecasting\\train 2.csv")
df

date  store  item  sales
0       2013-01-01      1     1     13
1       2013-01-02      1     1     11
2       2013-01-03      1     1     14
3       2013-01-04      1     1     13
4       2013-01-05      1     1     10
...            ...    ...   ...    ...
912995  2017-12-27     10    50     63
912996  2017-12-28     10    50     59
912997  2017-12-29     10    50     74
912998  2017-12-30     10    50     62
912999  2017-12-31     10    50     82

[913000 rows x 4 columns]

Exploratory Data Analysis(EDA)

In [3]:
df.nunique()

date     1826
store      10
item       50
sales     213
dtype: int64

findings: 
We can see that the data is clean of outliers and duplicates. Since, there will be 
1. 1826 days in 5 years period 
2. As per the data we have 10 unique stores and 50 unique products. 
3. We can not tell whether all the values in sales are right/wrong.

In [4]:
df.dtypes

date     object
store     int64
item      int64
sales     int64
dtype: object

Finding: We can see that all the columns are in 'int64' datatype other than date which is in 'object' datatype.

In [5]:
df['date'] = pd.to_datetime(df['date'])
df.dtypes

date     datetime64[ns]
store             int64
item              int64
sales             int64
dtype: object

Changed date column to 'datetime' format for better interpretation. 

In [6]:
print(df)

             date  store  item  sales
0      2013-01-01      1     1     13
1      2013-01-02      1     1     11
2      2013-01-03      1     1     14
3      2013-01-04      1     1     13
4      2013-01-05      1     1     10
...           ...    ...   ...    ...
912995 2017-12-27     10    50     63
912996 2017-12-28     10    50     59
912997 2017-12-29     10    50     74
912998 2017-12-30     10    50     62
912999 2017-12-31     10    50     82

[913000 rows x 4 columns]


In [7]:
x = df.groupby(['date','store','item']).sum()
x

sales
date       store item       
2013-01-01 1     1        13
                 2        33
                 3        15
                 4        10
                 5        11
...                      ...
2017-12-31 10    46       70
                 47       22
                 48       60
                 49       42
                 50       82

[913000 rows x 1 columns]

In [8]:
df2 = df.groupby(['date','item'], as_index = False).sales.sum()
df2

date  item  sales
0     2013-01-01     1    133
1     2013-01-01     2    327
2     2013-01-01     3    172
3     2013-01-01     4    102
4     2013-01-01     5     83
...          ...   ...    ...
91295 2017-12-31    46    605
91296 2017-12-31    47    211
91297 2017-12-31    48    494
91298 2017-12-31    49    285
91299 2017-12-31    50    626

[91300 rows x 3 columns]

In [9]:
df3 = df2.groupby(['item','date'],as_index = False).sales.sum()

In [10]:
df3

item       date  sales
0         1 2013-01-01    133
1         1 2013-01-02     99
2         1 2013-01-03    127
3         1 2013-01-04    145
4         1 2013-01-05    149
...     ...        ...    ...
91295    50 2017-12-27    511
91296    50 2017-12-28    587
91297    50 2017-12-29    596
91298    50 2017-12-30    612
91299    50 2017-12-31    626

[91300 rows x 3 columns]

In [11]:
df3['demand'] = df2.sales.rolling(window = 90).sum()

In [12]:
df3

item       date  sales   demand
0         1 2013-01-01    133      NaN
1         1 2013-01-02     99      NaN
2         1 2013-01-03    127      NaN
3         1 2013-01-04    145      NaN
4         1 2013-01-05    149      NaN
...     ...        ...    ...      ...
91295    50 2017-12-27    511  47760.0
91296    50 2017-12-28    587  47362.0
91297    50 2017-12-29    596  47072.0
91298    50 2017-12-30    612  46865.0
91299    50 2017-12-31    626  46793.0

[91300 rows x 4 columns]

In [13]:
df3.iloc[89:]

item       date  sales   demand
89        1 2013-03-31    203  24956.0
90        1 2013-04-01    134  24955.0
91        1 2013-04-02    187  24843.0
92        1 2013-04-03    210  24926.0
93        1 2013-04-04    229  24989.0
...     ...        ...    ...      ...
91295    50 2017-12-27    511  47760.0
91296    50 2017-12-28    587  47362.0
91297    50 2017-12-29    596  47072.0
91298    50 2017-12-30    612  46865.0
91299    50 2017-12-31    626  46793.0

[91211 rows x 4 columns]